### ISSUE TITLE
React-based views don't correctly update features

### ISSUE THREAD

**Issue by fregante** [refined-github/refined-github#6554]
I'm opening this issue to track it properly. In short:

1. Open https://github.com/refined-github/refined-github/blob/main/source/features/action-used-by-link.tsx
2. Visit a folder or change file via the sidebar or go back and forth

### Known issues

- `turbo:render` isn't called
- `popstate` events don't trigger any events
- the view isn't generated from scratch, so previous elements stick around 

### Related issues 

- #6414 
- #6528 

### Current status

For the whole existence of Refined GitHub, we've relied on _GitHub_ to delete our elements when the page/URL changes, with one exception (I don't remember which). New React-based views break this assumption. I think we need to:

- listen to `soft-nav:render`
- listen to `popstate` and avoid duplicates calls in case `soft-nav:render` is _also_ fired
- figure out a way to make widgets/buttons dynamic so that they change or disappear when the URL changes
	- this could help, but that alone doesn't fit into our concept of "feature unloading when pages change": #6493 
	- maybe we could add a `onAbort(() => element.remove())`

**Comment by 134130** [comment]
I usually used `MutationObserver`.
I think watching popstate of Element is more proper to use It

**Comment by fregante** [comment]
> I think using It is more proper for watching popstate of Element.

Each page/view is still tracked "by URL", so listening to URL changes is definitely the proper way to do it.

Detecting changes isn't the problem, our problem is that we need to respond to those changes.

We can already use our selector observer to watch the page for specific elements without tracking "every change". The only disadvantage is that it doesn't tell us when elements are removed, but we actually have a separate listener specifically for that: `onElementRemoval` and `onElementReplacement` (which haven't really been needed since we added the selector observer)

In Refined GitHub we have a global feature manager that loads and unloads features based on a "page change". There is no easy and reliable way to detect a "page change" with a MutationObserver because some elements are dropped/regenerated by React and that doesn't mean that the page was updated.

**Comment by fregante** [comment]
There might be a way to reliabley catch URL changes:

- https://github.com/fregante/webext-events/issues/6

However I don't know yet:

- are the events fast enough to replace `turbo:render` and avoid FOUC? It seems to me that they're actually early (in the background page) but there could be some delay when passing the message to the content script
- does RGH react well when a "url change" event is received but the page doesn't change? e.g. pressing `y` on files will update just the URL
- could this event be early enough to be used as an "unload" event?
	- #6437  

Unresolved issues:

- when the page changes, but the DOM is only partially updated, which is causing #6414

**Comment by fregante** [comment]
In Chrome we can already use the navigation API 

https://developer.mozilla.org/en-US/docs/Web/API/Navigation/navigate_event

**Comment by fregante** [comment]
Nope. It fires before the actual navigation, so we'd still have to wait for the content to be loaded somehow.

However I think I have a solution: `.turbo-progress-bar`

https://github.com/hotwired/turbo/blob/c44664d6acea224502d440754e2376635a6e9652/src/core/drive/progress_bar.js

- This element is added to the page on click and **removed on load**
- the element appears between `head` and `body`
- it works on folders and files
- the same element is reused

Detecting this removal might be a bit complicated, or maybe we just need a good ol MutationObserver on `html`, which should be lightweight

**Comment by fregante** [comment]
I solved the URL detection issue in the PR above, but that's only part of it.

The main issue is that the DOM is no longer actually re-generated:

1. our selector observer tags a "seen" element
2. after a page change the element is never removed
3. no `add` callback is ever fired

In practice this means our features are not updated. I think we just need dynamic elements:

1. the feature is initialized with the current domloaded/`turbo:load` event
2. use the selector observer to attach a React-ish element
3. use the MutationObserver-based listener to detect changes to the view
4. update the React-ish element 

This brings us back to discussions like #6493, where we need to find a usable library and then work the new MutationObserver-based listener into it.

**Comment by 134130** [comment]
> The main issue is that the DOM is no longer actually re-generated:
> 
> 1. our selector observer tags a "seen" element
> 2. the contents page, but the element is never removed
> 3. no add callback is ever fired

selector observer generates "seen" mark with caller id.
How about adding additional identifier which is based on `location.href` to "seen" mark?

If we can detect path changing correctly and ensure that feature is based on `location.href`, It will be easier and seamless way to support.

**Comment by fregante** [comment]
That's not going to be enough because we also need to remove the old element from the page. At that point it's easier to write:

- on page unload, remove `seen` and element

Then the observer will take care of the rest, no need to add a dependency to the URL.

Maybe that's just what we need though:

- have `selector-observer` remove the `seen` class when its `signal` is aborted
- also add `onAbort(signal, () => element.remove())` to all of these features 

It could be a start and it wouldn't require writing dynamic elements.

**Comment by ilike2burnthing** [comment]
Before I open a new issue, would this include the following?

- go to https://github.com/refined-github/refined-github/tree/default-button-fixes
- click `default-branch-button`
- be returned to main branch but `default-branch-button` still remains, until page is refreshed or button is clicked again
- if page isn't refreshed or button clicked again, then clicking `Code` tab at the top left  still leads to https://github.com/refined-github/refined-github/tree/default-button-fixes

**Comment by fregante** [comment]
All the features around files and file lists are broken (`/blob/` and `/tree/`, not `/pull/`))

**Comment by fregante** [comment]
My latest attempt in https://github.com/refined-github/refined-github/pull/7539#discussion_r1672360261

In short:

1. load the feature as usual
2. add a React view-specific page load listener to update the visibility and data of the widget

This works reasonably well, with the exception that include/excludes are not run at this point, so if the first page load is not in `include`, successive ones will not enable the feature.

**Comment by SunsetTechuila** [comment]
Another perspective: we could take the same approach as Discord mods and patch components directly instead of manipulating the DOM. Vencord exists as an extension, so this should be doable

**Comment by SunsetTechuila** [comment]
> listen to soft-nav:render

What is preventing us from doing this right now? I didn't notice new bugs after replacing `turbo:render` with `soft-nav:render` in [feature-manager.tsx](https://github.com/refined-github/refined-github/blob/80c004a03187a59c0f46e94b5dce1905b173c6ef/source/feature-manager.tsx), only some of the current ones disappeared

**Comment by fregante** [comment]
This issue is resolved as of v26.7.26. Some features are still affected but for the most part I think we're good. Feature-specific issues can be opened separately.
